File: src/publisher.cpp
//
// Author: Ori Novanda (cargmax-at-gmail.com)
//

#include "ros/ros.h"
#include "performance_tests/SuperAwesome.h"
#include "performance_tests/SetRate.h"
#include <sstream>

double sendRate = 10;
bool sendRateUpdated = false;

bool setRate(performance_tests::SetRate::Request &req,
		performance_tests::SetRate::Response &res) {
	sendRate = req.rate;
	if(sendRate != 0) {
		//rate = ros::Rate(sendRate);
		ROS_INFO("Publish rate: %lf", sendRate);
		sendRateUpdated = true;
	}
	else {
		ROS_INFO("Pause sending");
	}
	return true;	
}
		
int main(int argc, char **argv) {
	ros::Time::init();
	ros::init(argc, argv, "c_publisher");
	ros::NodeHandle nh;
        ros::Publisher pub = nh.advertise<performance_tests::SuperAwesome>("benchmark",100);
	ros::ServiceServer svc =  nh.advertiseService("SetPublishRate", setRate);
	
	ros::Rate rate = ros::Rate(sendRate);

	int count = 1;
	performance_tests::SuperAwesome msg;
	while (ros::ok()) {
		if(sendRate == 0) {
			ros::spinOnce();
			ros::Duration(0.1).sleep();
			continue;
		}
		else if(sendRateUpdated) {
			rate = ros::Rate(sendRate);
			sendRateUpdated = false;
		}
		std::stringstream ss;
		ss << count++;
		msg.data = ss.str();
		//ROS_INFO("%s", msg.data.c_str());
		pub.publish(msg);
		ros::spinOnce();
		if(sendRate == 0) continue;
		rate.sleep();
	}
	return 0;
}

File: scripts/publisher.py
#!/usr/bin/env python

#
# Author: Ori Novanda (cargmax-at-gmail.com)
#

import rospy, time
from performance_tests.msg import SuperAwesome
from performance_tests.srv import *

sendRate = 10
rate = None

def setRate_Handler(req):
	global rate, sendRate
	sendRate = req.rate
	if sendRate != 0:
		rate = rospy.Rate(sendRate)
		rospy.loginfo("Publish rate: %f", req.rate)
	else:
		rospy.loginfo("Pause sending")
	return SetRateResponse()

def publisher():
	global rate
	pub = rospy.Publisher("benchmark", SuperAwesome, queue_size=100)
	rospy.init_node("python_publisher", anonymous=True)
	s = rospy.Service("SetPublishRate", SetRate, setRate_Handler)
	rate = rospy.Rate(sendRate)
	count = 1
	msg = SuperAwesome()
	while not rospy.is_shutdown():
		if rate == 0:
			time.sleep(0.1)
			next
		msg.data = "%d" % count;
		#rospy.loginfo(msg)
		pub.publish(msg)
		count += 1
		if sendRate == 0:
			next
		rate.sleep()

if __name__ == "__main__":
	try:
		publisher()
	except rospy.ROSInterruptException:
		pass

